src/MDOP.jl

```diff
@@ -4,16 +4,150 @@ using Distances
+FNN-statistic `FNNs` increases . The final embedding vector is stored in `Y`
+(`Dataset`). The chosen delay values for each embedding cycle are stored in the
+`τ_vals` and the according time series number chosen for the according delay
+value in `τ_vals` is stored in `ts_vals`. For univariate embedding (`s::Vector`)
+`ts_vals` is a vector of ones of length `τ_vals`, because there is simply just
+one `β`-statistic for each embedding cycle as an `Array` of `Vector`s.
 
 [^Nichkawde2013]: Nichkawde, Chetan (2013). [Optimal state-space reconstruction using derivatives on projected manifold. Physical Review E 87, 022905](https://doi.org/10.1103/PhysRevE.87.022905).
+[^Hegger1999]: Hegger, Rainer and Kantz, Holger (1999). [Improved false nearest neighbor method to detect determinism in time series data. Physical Review E 60, 4970](https://doi.org/10.1103/PhysRevE.60.4970).
+[^Kennel1992]: Kennel, M. B., Brown, R., Abarbanel, H. D. I. (1992). [Determining embedding dimension for state-space reconstruction using a geometrical construction. Phys. Rev. A 45, 3403] (https://doi.org/10.1103/PhysRevA.45.3403).
 """
-function MDOP(Y...)
+function MDOP(s::Vector{<:Real}; τs = 0:50 , w::Int = 1, fnn_thres::Real = 0.05,
+    r::Real = 2, β_condition::Bool=false)
+    @assert 0 <= fnn_thres < 1 "Please select a valid breaking criterion, i.e. a threshold value `fnn_thres` ∈ [0 1)"
+    @assert all(x -> x ≥ 0, τs)
+
+    T = eltype(s)
+
+    max_num_of_cycles = 50 # assure that the algorithm will break after 50 embedding cycles
+
+    metric = Euclidean()
+    # normalize input time series (especially important for fnn-computation)
+    s .= (s.-mean(s))./std(s)
+    # define actual phase space trajectory
+    Y_act = Dataset(s)
+
+    # compute nearest neighbor distances in the first embedding dimension for
+    # FNN statistic
+    vtree = KDTree(Y_act, metric)
+    allNNidxs, NNdist_old = all_neighbors(vtree, Y_act, 1:length(Y_act), 1, w)
+
+    # set a flag, in order to tell the while loop when to stop. Each loop
+    # stands for encountering a new embedding dimension
+    flag = true
+
+    # set index-counter for the while loop
+    cnt = 1;
+
+    # preallocate output variables
+    τ_vals = zeros(Int,1)
+    ts_vals = ones(Int,1)
+    FNNs = zeros(Float64,1)
+    if β_condition; βS = Array{T}(undef, length(τs), max_num_of_cycles); end
+
+    # loop over increasing embedding dimensions until some break criterion will
+    # tell the loop to stop/break
+    while flag
+
+        # get the β-statistic
+        β = beta_statistic(Y_act, s; τs = τs , w = w)
+        if β_condition; βS[:, cnt] = β; end
+
+        # determine the optimal tau value from the β-statistic
+        maxi, max_idx = findmax(β)
+        # store chosen delay (and chosen time series)
+        push!(τ_vals, τs[max_idx])
+        push!(ts_vals, 1)
+
+        # create phase space vector for this embedding cycle
+        Y_act = DelayEmbeddings.embed_one_cycle(Y_act,s,τ_vals[cnt+1])
```

Comment:
Notice that this is a type-unstable operation. The dimensionality of a dataset is part of its type. Here you replace `Y_act` with a version of itself with different dimensionality and thus different type.

However, with this algorithm design, there isn't a better alterantive and we will have just ignore this type instability and focus on doing proper profiling instead to see if it affects anything.